projects/sandbox/train/train/data_structures.py

```diff
@@ -76,12 +76,47 @@ def __next__(self) -> Tuple[torch.Tensor, torch.Tensor]:
 
 
 class BBHNetWaveformInjection(RandomWaveformInjection):
+    def __init__(self, *args, **kwargs):
+        try:
+            glitch_prob = kwargs.pop("glitch_prob")
+            downweight = kwargs.pop("downweight")
+            prob = kwargs.pop("prob")
+        except KeyError:
+            self.downweight = 1
+        else:
+            # not to sound like Fermat but I have a
+            # derivation of this somewhere that I
+            # can't quite find at the moment
+            prob = prob / (1 - glitch_prob * (1 - downweight)) ** 2
```

Comment:
This step is so that we still have the same expected number of waveforms?

Reply:
Exactly. If you let `p` be the desired expected number of waveforms and `p_0` be the new probability for samples that _don't_ have a glitch inserted, with `d` being the downweight factor and `g` being the glitch insertion probability, you want

$$p = p_0(1 - g)^2 + 2(p_0d)g(1-g) + (p_0d^2)g^2$$

$$p = p_0((1 - g)^2 + 2(1 - g)(dg) + (dg)^2)$$

$$p = p_0((1-g) + gd)^2 = p_0(1 + g(d - 1))^2$$

$$p_0 = \frac{p}{(1 + g(d - 1))^2}$$

Reply:
Nice, this all makes sense then